data exporting support, supports exporting task history via csv

snicker opened the pull request:
- adds the following new dependencies: 
  - express-csv
- recommend replacing functionality of 'relative-date' with features from 'moment'
- adds new routes for data export
- supports retrieval of CSV history via the API and in browser
  - /api/v1/export/history (requires typical API authorization)
  - /export/history.csv (requires session authorization)
- adds new "My Data" tab to the profile page with a link to the CSV

snicker committed:
data export commit, supports exporting task history via csv
*adds the following new dependencies: moment, express-csv
*recommend replacing functionality of 'relative-date' with features from 'moment'
*supports retrieval of CSV history via the API and in browser
**/api/v1/export/history (requires typical API authorization)
**/export/history.csv (requires session authorization)
*adds new routes for data export

snicker committed:
Merge branch 'develop' of https://github.com/snicker/habitrpg into develop

snicker commented:
This should eventually be followed up with tests, code will probably need a modest amount of refactoring to be more testable

paglias commented:
Moment is already a dependency, at the beginning of package.json dependencies

snicker committed:
derp. moment is already a dependency

snicker commented:
thanks @paglias, I completely missed that. I think I coded some of this half asleep.

snicker committed:
ack, trailing comma

paglias commented:
There's in some old issue a request to give the possibility to see all the actions made by the user? I'm from phone so I'm not sure but the title should be similar to "Log activities to view later". 

Your code could be adapted to show all the activities on the site and not only downloading them? We should have a discussion about this feature first but I think it's a good starting point.

Also I'm not completely sure about this but couldn't the auth method from the auth controller be used as a middleware to ensure the user is logged in? Give a look at src/routes/api.js to see how it's being used

paglias commented:
Sorry about the last part of my comment, you're using auth.auth.

What is dataexport.auth used for then?

snicker commented on src/controllers/dataexport.js:
```diff
@@ -0,0 +1,51 @@
+
+dataexport.auth = function(req, res, next) { //[todo] there is probably a more elegant way of doing this...
+  var uid;
+  uid = req.session.userId;
```
`auth.auth` relies on the client passing in their user id and api key in the header to authenticate. This is great for ajax calls, but won't work when we need to download files, so we need a way for the router to authenticate via the current session. That is all `dataexport.auth` does. As I mentioned above, there is probably a more elegant way of doing that, but I couldn't immediately figure it out. I'm guessing something to do with `passport` and loading the current session, but no such function exists for the express router. Perhaps one should be added to the auth controller.

paglias replied:
Mmm... Ok, i had to use a similar approach for translations, we should add an auth method that doesn't rely on http headers somewhere so that we don't duplicate code or add support for passing auth parameters as query parameter (but that wouldn't work for translations when serving the index jade file)

snicker replied:
Do users need to be authed to get translation data? Shouldn't that
kind of stuff be statically served? Forgive me, I'm not too familiar
with what you're doing :) but I agree that it should probably be moved
to the `Auth` controller, maybe `Auth.authWithSession`

paglias replied:
Yeah no need to be authorized but we need to access their profile in order to get their `preferences.language`

snicker commented:
re: "Log activities to view later", it's definitely easy to change this code into that. It should be refactored anyway to provide a method that generates an array of history from a particular user which is called from the export methods (This gives us the opportunity to easily add more export formats, XML, RTF, TXT, XLS, etc), which can then be used anywhere. I'd be inclined to call such a controller "data" or something similar

lefnire commented:
Oh this is so cool! You're gonna make a lot of users happy. I'm gonna start toying with it and get back to you.

Quick UI thing - thoughts on putting "export" button inside settings tab, since it's just one button? Or are there planned "export options" for future?

Also, @snicker ol' buddy ol' pal :) After reading the auth.auth / exports.auth convo (an issue that vexed me way back when, and I didn't know a solution) - You might be the perfect candidate for getting the [old iCal code](https://github.com/HabitRPG/habitrpg/blob/develop/src/controllers/deprecated.js#L37) working with new site :) :)

snicker commented:
I was thinking that the export stuff should go in the settings tab as
well, but it needs to be a little further tab-ified, and I didn't
wanna add too much UI muck with this. But yes, I envision some more
export options going somewhere, so I could make these couple changes
and do another push before merging

snicker committed:
moved UI stuff into the settings area, moved authWithSession to the Auth controller

snicker commented:
There we go, moved everything to the settings area, made the UI a little more verbose, and moved the authorization stuff to the `Auth` controller. Merge away
